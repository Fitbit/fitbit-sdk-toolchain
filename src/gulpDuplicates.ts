import stream from 'stream';

import humanizeList from 'humanize-list';
import PluginError from 'plugin-error';
import Vinyl from 'vinyl';

const PLUGIN_NAME = 'duplicates';

export default function vinylAssertFiles() {
  const foundPaths = new Set<string>();
  const duplicatePaths = new Set<string>();
  return new stream.Transform({
    objectMode: true,
    transform(this: stream.Transform, file: Vinyl, _, cb) {
      if (!file.isNull()) {
        if (foundPaths.has(file.relative)) {
          duplicatePaths.add(file.relative);
        } else {
          foundPaths.add(file.relative);
        }
      }
      cb(undefined, file);
    },
    flush(this: stream.Transform, cb) {
      if (duplicatePaths.size !== 0) {
        this.emit(
          'error',
          new PluginError(
            PLUGIN_NAME,
            `One or more files in the build output were duplicates: ${humanizeList(
              [...duplicatePaths],
            )}`,
          ),
        );
      }
      cb();
    },
  });
}
