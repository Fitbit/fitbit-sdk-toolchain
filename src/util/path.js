export function dirname(path) {
  let dir = path.split('/').slice(0, -1).join('/');
  if (dir === '') dir = '/';
  return dir;
}

export function basename(path) {
  return path.replace(/\\/g, '/').replace(/.*\//, '');
}

export function split(path) {
  return [dirname(path), basename(path)];
}
