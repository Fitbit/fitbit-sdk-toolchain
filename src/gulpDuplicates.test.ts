import { Readable } from 'stream';
import path from 'path';

import Vinyl from 'vinyl';
import gulpDuplicates from './gulpDuplicates';

function makeTestStream(paths: string[]) {
  const stream = new Readable({ objectMode: true });
  stream._read = () => {};

  for (const path of paths) {
    stream.push(new Vinyl({ path, contents: Buffer.alloc(0) }));
  }

  stream.push(null);
  return stream;
}

function expectStreamPromise(paths: string[]) {
  return expect(
    new Promise((resolve, reject) => {
      const stream = makeTestStream(paths).pipe(gulpDuplicates());
      stream.on('error', reject);
      stream.on('finish', resolve);
    }),
  );
}

it('does not emit an error when no files are duplicated', () =>
  expectStreamPromise([
    'app/index.js',
    'app/foo.js',
    'resources/icon.png.txi',
  ]).resolves.toBeUndefined());

it('emits an error when a file is duplicated twice', () =>
  expectStreamPromise(['app/index.js', 'app/index.js']).rejects.toThrowError(
    `One or more files in the build output were duplicates: ${path.join(
      'app',
      'index.js',
    )}`,
  ));

it('emits an error when two files are duplicated', () =>
  expectStreamPromise([
    'app/index.js',
    'app/index.js',
    'resources/icon.png.txi',
    'resources/icon.png.txi',
  ]).rejects.toThrowError(
    `One or more files in the build output were duplicates: ${path.join(
      'app',
      'index.js',
    )} and ${path.join('resources', 'icon.png.txi')}`,
  ));
