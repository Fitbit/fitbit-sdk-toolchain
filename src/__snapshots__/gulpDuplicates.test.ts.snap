// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emits an error when a file is duplicated twice 1`] = `"One or more files in the build output were duplicates: app/index.js"`;

exports[`emits an error when two files are duplicated 1`] = `"One or more files in the build output were duplicates: app/index.js and resources/icon.png.txi"`;
